File: create-buckets.js
import { S3Client, CreateBucketCommand, PutBucketCorsCommand, PutBucketEncryptionCommand, PutBucketVersioningCommand, PutBucketPolicyCommand, HeadBucketCommand } from '@aws-sdk/client-s3';
import dotenv from 'dotenv';

// Load environment variables
dotenv.config();

// Initialize S3 client
const s3Client = new S3Client({
  region: process.env.VITE_AWS_REGION,
  credentials: {
    accessKeyId: process.env.VITE_AWS_ACCESS_KEY_ID,
    secretAccessKey: process.env.VITE_AWS_SECRET_ACCESS_KEY
  }
});

// Check if bucket exists
async function bucketExists(bucketName) {
  try {
    await s3Client.send(new HeadBucketCommand({ Bucket: bucketName }));
    return true;
  } catch (error) {
    if (error.$metadata?.httpStatusCode === 404) {
      return false;
    }
    throw error;
  }
}

// Create and configure bucket
async function createAndConfigureBucket(bucketName) {
  try {
    // Check if bucket exists
    const exists = await bucketExists(bucketName);
    if (exists) {
      console.log(`Bucket ${bucketName} already exists`);
      return true;
    }

    // Create bucket
    await s3Client.send(new CreateBucketCommand({
      Bucket: bucketName,
      CreateBucketConfiguration: {
        LocationConstraint: process.env.VITE_AWS_REGION === 'us-east-1' ? undefined : process.env.VITE_AWS_REGION
      }
    }));
    console.log(`Created bucket: ${bucketName}`);

    // Configure CORS
    await s3Client.send(new PutBucketCorsCommand({
      Bucket: bucketName,
      CORSConfiguration: {
        CORSRules: [{
          AllowedHeaders: ['*'],
          AllowedMethods: ['GET', 'PUT', 'POST', 'DELETE'],
          AllowedOrigins: ['*'],
          ExposeHeaders: ['ETag'],
          MaxAgeSeconds: 3000
        }]
      }
    }));

    // Configure encryption
    await s3Client.send(new PutBucketEncryptionCommand({
      Bucket: bucketName,
      ServerSideEncryptionConfiguration: {
        Rules: [{
          ApplyServerSideEncryptionByDefault: {
            SSEAlgorithm: 'AES256'
          }
        }]
      }
    }));

    // Enable versioning
    await s3Client.send(new PutBucketVersioningCommand({
      Bucket: bucketName,
      VersioningConfiguration: {
        Status: 'Enabled'
      }
    }));

    // Set bucket policy
    const bucketPolicy = {
      Version: '2012-10-17',
      Statement: [{
        Sid: 'AllowAuthenticatedAccess',
        Effect: 'Allow',
        Principal: '*',
        Action: [
          's3:GetObject',
          's3:PutObject',
          's3:DeleteObject'
        ],
        Resource: [
          `arn:aws:s3:::${bucketName}`,
          `arn:aws:s3:::${bucketName}/*`
        ]
      }]
    };

    await s3Client.send(new PutBucketPolicyCommand({
      Bucket: bucketName,
      Policy: JSON.stringify(bucketPolicy)
    }));

    console.log(`Successfully configured bucket: ${bucketName}`);
    return true;
  } catch (error) {
    console.error(`Error creating/configuring bucket ${bucketName}:`, error);
    return false;
  }
}

// Main function
async function main() {
  // Validate environment variables
  const requiredVars = [
    'VITE_AWS_REGION',
    'VITE_AWS_ACCESS_KEY_ID',
    'VITE_AWS_SECRET_ACCESS_KEY',
    'VITE_AWS_BUCKET_NAME'
  ];

  const missingVars = requiredVars.filter(varName => !process.env[varName]);
  if (missingVars.length > 0) {
    console.error('Missing required environment variables:', missingVars.join(', '));
    process.exit(1);
  }

  const baseBucketName = process.env.VITE_AWS_BUCKET_NAME;
  const suffixes = ['raw', 'processed', 'embeddings'];

  console.log('Creating and configuring buckets...');
  
  for (const suffix of suffixes) {
    const bucketName = `${baseBucketName}-${suffix}`;
    await createAndConfigureBucket(bucketName);
  }
}

// Run the script
main().catch(error => {
  console.error('Failed to create buckets:', error);
  process.exit(1);
});

File: create_buckets.py
import os
import boto3
from botocore.exceptions import ClientError

def create_bucket(s3_client, bucket_name, region):
    """Create an S3 bucket with configuration"""
    try:
        location_constraint = None if region == 'us-east-1' else region
        s3_client.create_bucket(
            Bucket=bucket_name,
            CreateBucketConfiguration={'LocationConstraint': location_constraint} if location_constraint else {}
        )
        print(f"Created bucket: {bucket_name}")

        # Configure CORS
        s3_client.put_bucket_cors(
            Bucket=bucket_name,
            CORSConfiguration={
                'CORSRules': [{
                    'AllowedHeaders': ['*'],
                    'AllowedMethods': ['GET', 'PUT', 'POST', 'DELETE'],
                    'AllowedOrigins': ['*'],
                    'ExposeHeaders': ['ETag'],
                    'MaxAgeSeconds': 3000
                }]
            }
        )

        # Configure encryption
        s3_client.put_bucket_encryption(
            Bucket=bucket_name,
            ServerSideEncryptionConfiguration={
                'Rules': [{
                    'ApplyServerSideEncryptionByDefault': {
                        'SSEAlgorithm': 'AES256'
                    }
                }]
            }
        )

        # Enable versioning
        s3_client.put_bucket_versioning(
            Bucket=bucket_name,
            VersioningConfiguration={'Status': 'Enabled'}
        )

        # Set bucket policy
        bucket_policy = {
            'Version': '2012-10-17',
            'Statement': [{
                'Sid': 'AllowAuthenticatedAccess',
                'Effect': 'Allow',
                'Principal': '*',
                'Action': ['s3:GetObject', 's3:PutObject', 's3:DeleteObject'],
                'Resource': [f'arn:aws:s3:::{bucket_name}/*']
            }]
        }
        s3_client.put_bucket_policy(
            Bucket=bucket_name,
            Policy=str(bucket_policy).replace("'", '"')
        )

        print(f"Configured bucket {bucket_name} successfully")
        return True

    except ClientError as e:
        error_code = e.response.get('Error', {}).get('Code')
        if error_code == 'BucketAlreadyOwnedByYou':
            print(f"Bucket {bucket_name} already exists and is owned by you")
            return True
        elif error_code == 'BucketAlreadyExists':
            print(f"Bucket {bucket_name} already exists but is owned by another account")
        else:
            print(f"Error creating bucket {bucket_name}: {str(e)}")
        return False

def main():
    # Get AWS credentials from environment
    region = os.getenv('VITE_AWS_REGION')
    access_key = os.getenv('VITE_AWS_ACCESS_KEY_ID')
    secret_key = os.getenv('VITE_AWS_SECRET_ACCESS_KEY')
    base_bucket_name = os.getenv('VITE_AWS_BUCKET_NAME')

    if not all([region, access_key, secret_key, base_bucket_name]):
        print("Missing required environment variables")
        return

    # Initialize S3 client
    s3_client = boto3.client(
        's3',
        region_name=region,
        aws_access_key_id=access_key,
        aws_secret_access_key=secret_key
    )

    # Create buckets
    suffixes = ['raw', 'processed', 'embeddings']
    success = True
    for suffix in suffixes:
        bucket_name = f"{base_bucket_name}-{suffix}"
        if not create_bucket(s3_client, bucket_name, region):
            success = False
            break

    if success:
        print("All buckets created and configured successfully!")
    else:
        print("Failed to create all buckets")

if __name__ == '__main__':
    main()